### PR TITLE
feat: Implement infinite approval (SC-5372)

### PR DESCRIPTION
**Pull request by edag94**
- No change to allowance with infinite approval

### CODE REVIEW

File: contracts/test/ERC20.t.sol [hunk 1]
```diff
@@ -124,6 +139,30 @@ contract ERC20BaseTest is TestUtils {
         }
     }
 
+    function testFuzz_transferFrom_infiniteApproval(address recipient_, uint256 amount_) public {
+        amount_                  = constrictToRange(amount_,   0, type(uint256).max);
+        uint256 infiniteApproval = type(uint256).max;
+        if (amount_ > infiniteApproval) return;  // Owner must approve for more than amount.
```
Comment (vbidin) [inline contracts/test/ERC20.t.sol]: When would this condition trigger?

File: contracts/test/ERC20.t.sol [hunk 2]
```diff
@@ -124,6 +139,30 @@ contract ERC20BaseTest is TestUtils {
...
+        owner.erc20_approve(address(_token), self, infiniteApproval);
+
+        assertTrue(_token.transferFrom(address(owner), recipient_, amount_));
+
+        assertEq(_token.totalSupply(), amount_);
+
+        assertEq(_token.allowance(address(owner), self), infiniteApproval);
+
+        if (address(owner) == recipient_) {
+            assertEq(_token.balanceOf(address(owner)), amount_);
```
Comment (vbidin) [inline contracts/test/ERC20.t.sol:162]: You can move this check out of the if condition since it's identical in both cases.

**Review by vbidin — DISMISSED** [review]

File: contracts/test/ERC20.t.sol (hunk 1 above)
Comment (edag94) [reply contracts/test/ERC20.t.sol]: Never, this was accidently left in when fixing the test

File: contracts/ERC20.sol [hunk 3]
```diff
@@ -145,6 +146,11 @@ contract ERC20 is IERC20 {
         emit Transfer(owner_, address(0), amount_);
     }
 
+    function _decreaseAllowance(address owner_, address spender_, uint256 subtractedAmount_) internal {
+        uint256 spenderAllowance = allowance[owner_][spender_];  // Cache to memory.
+        if (spenderAllowance != type(uint256).max) _approve(owner_, spender_, spenderAllowance - subtractedAmount_);
```
Comment (lucas-manuel) [inline contracts/ERC20.sol]:
```js
if (spenderAllowance != type(uint256).max) {
    _approve(owner_, spender_, spenderAllowance - subtractedAmount_);
}
```

File: contracts/test/ERC20.t.sol (hunk 1 above)
Comment (lucas-manuel) [reply contracts/test/ERC20.t.sol]: Lets remove this line

File: contracts/test/ERC20.t.sol [hunk 4]
```diff
@@ -124,6 +139,30 @@ contract ERC20BaseTest is TestUtils {
...
+        } else {
+            assertEq(_token.balanceOf(address(owner)), 0);
+            assertEq(_token.balanceOf(recipient_), amount_);
```
Comment (lucas-manuel) [inline contracts/test/ERC20.t.sol]: Alignment

File: contracts/test/ERC20.t.sol [hunk 5]
```diff
@@ -124,6 +139,30 @@ contract ERC20BaseTest is TestUtils {
...
+        ERC20User owner = new ERC20User();
+
+        _token.mint(address(owner), amount_);
+        owner.erc20_approve(address(_token), self, infiniteApproval);
+
+        assertTrue(_token.transferFrom(address(owner), recipient_, amount_));
+
+        assertEq(_token.totalSupply(), amount_);
+
+        assertEq(_token.allowance(address(owner), self), infiniteApproval);
```
Comment (lucas-manuel) [inline contracts/test/ERC20.t.sol]: Add remove line break and align

File: contracts/test/ERC20.t.sol [hunk 6]
```diff
@@ -101,6 +114,8 @@ contract ERC20BaseTest is TestUtils {
     }
 
     function testFuzz_transferFrom(address recipient_, uint256 approval_, uint256 amount_) public {
+        approval_ = constrictToRange(approval_, 0, type(uint256).max - 1);
+        amount_   = constrictToRange(amount_,   0, type(uint256).max - 1);
         if (amount_ > approval_) return;  // Owner must approve for more than amount.
```
Comment (lucas-manuel) [inline contracts/test/ERC20.t.sol]:
```js
approval_ = constrictToRange(approval_, 0, type(uint256).max - 1);
amount_   = constrictToRange(amount_,   0, approval_);
```

and can remove if statement

File: contracts/ERC20.sol [hunk 7]
```diff
@@ -63,8 +63,9 @@ contract ERC20 is IERC20 {
     }
 
     function decreaseAllowance(address spender_, uint256 subtractedAmount_) external override returns (bool success_) {
-        _approve(msg.sender, spender_, allowance[msg.sender][spender_] - subtractedAmount_);
+        _decreaseAllowance(msg.sender, spender_, subtractedAmount_);
         return true;
+
```
Comment (lucas-manuel) [inline contracts/ERC20.sol]: Remove extra line

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

**Pull request merged**